common/configuration/cache.go

```diff
@@ -0,0 +1,281 @@
+				continue
+			}
+			list, err := c.listProvider.List(ctx)
+			if err != nil {
+				logger.Warnf("could not sync: could not get list: %v", err)
+				continue
+			}
+			for _, cp := range providersToSync {
+				_, hasValues := slices.Find(list, func(e Entry) bool {
+					return ProviderKeyForAccessor(e.Accessor) == cp.provider.Key()
+				})
+				wg.Add(1)
+				go func(cp *cacheProvider[R]) {
+					cp.sync(ctx, clock, hasValues)
+					wg.Done()
+				}(cp)
+			}
+			wg.Wait()
+			next = clock.Now().Add(time.Second)
+		}
+	}
+}
+
+func (c *cache[R]) processEvent(e updateCacheEvent) {
+	if pv, ok := c.providers[e.key]; ok {
+		pv.processEvent(e)
+	}
+	// waitGroup updated so testing can wait for events to be processed
+	c.topicWaitGroup.Done()
+}
+
+// cacheProvider wraps an asynchronous provider and caches its values.
+type cacheProvider[R Role] struct {
+	provider AsynchronousProvider[R]
+	values   *xsync.MapOf[Ref, SyncedValue]
+
+	// isActive is the provider is used by any value
+	// When inactive, the provider will not be synced. This helps avoid accessing resources that are not needed, like 1Password.
+	// When the provider stops being used, we do one final sync to ensure the cache is up-to-date
+	isActive bool
+
+	loaded     chan bool  // closed when values have been synced for the first time
+	loadedOnce *sync.Once // ensures we close the loaded channel only once
+
+	lastSyncAttempt optional.Option[time.Time] // updated each time we attempt to sync, regardless of success/failure
+	currentBackoff  optional.Option[time.Duration]
+}
+
+// waitForInitialSync waits until the initial sync has completed.
+//
+// If values have not yet been synced, this will wait until they are, returning an error if it takes too long.
+func (c *cacheProvider[R]) waitForInitialSync() error {
+	select {
+	case <-c.loaded:
+		return nil
+	case <-time.After(waitForCacheTimeout):
+		return fmt.Errorf("%s has not completed sync yet", c.provider.Key())
+	}
+}
+
+// needsSync returns true if the provider needs to be synced.
+func (c *cacheProvider[R]) needsSync(clock clock.Clock) bool {
+	lastSyncAttempt, ok := c.lastSyncAttempt.Get()
+	if !ok {
+		return true
+	}
+	if currentBackoff, ok := c.currentBackoff.Get(); ok {
+		return clock.Now().After(lastSyncAttempt.Add(currentBackoff))
+	}
+	return clock.Now().After(lastSyncAttempt.Add(c.provider.SyncInterval()))
+}
+
+// sync executes sync on the provider and updates the cacheProvider sync state
+func (c *cacheProvider[R]) sync(ctx context.Context, clock clock.Clock, hasValues bool) {
+	logger := log.FromContext(ctx)
+
+	if !hasValues && !c.isActive {
+		// skip
+		c.loadedOnce.Do(func() {
+			// treat this as a successful initial sync, so we don't ever block trying to access this provider
+			close(c.loaded)
+		})
+		return
+	}
+
+	c.lastSyncAttempt = optional.Some(clock.Now())
+	err := c.provider.Sync(ctx, c.values)
+	if err != nil {
+		logger.Errorf(err, "Error syncing %s", c.provider.Key())
```

Comment:
This happens when a provider fails? eg asm timeout? Does the backoff time out?

Reply:
This happens when a provider fails to sync (such as ASM timing out, or 1Password not getting authorized).
The backoff schedules the sync to try again in a future round of async provider syncs, starting at 1s later, exponentially growing and then maxing out at a 2 min maximum (`syncMaxBackoff`).
Does that answer your question @gak?